### PR TITLE
Added more missing annotations to Typescript definition file

### PR DESCRIPTION
**Pull request by arusanov**
1. Changed `ref` argument to `Component`
2. Added `defaultValue` for inputs
3. Added `context` property to `Component`

### CODE REVIEW

File: src/preact.d.ts [hunk 1]
```diff
@@ -11,7 +11,8 @@ declare namespace preact {
 	interface PreactHTMLAttributes {
 		dangerouslySetInnerHTML?:DangerouslySetInnerHTML;
 		key?:string;
-		ref?:(el?: Element) => void;
+		defaultValue?:string;
+		ref?:<T,S>(component:Component<T,S>) => void;
```
Comment (developit) [inline src/preact.d.ts]: `ref` can receive both components and elements.

Comment (arusanov) [inline src/preact.d.ts]: fixed in https://github.com/developit/preact/pull/350/commits/791cfd4c0195c5cce872fa235da8386054b0db7e

**Comment by developit** [comment]
LGTM but I've basically never used TypeScript.
Maybe @jrf0110 or @kruczy can confirm 😀

File: src/preact.d.ts [hunk 2]
```diff
@@ -11,7 +11,8 @@ declare namespace preact {
 	interface PreactHTMLAttributes {
 		dangerouslySetInnerHTML?:DangerouslySetInnerHTML;
 		key?:string;
-		ref?:(el?: Element) => void;
+		defaultValue?:string;
```
Comment (kruczy) [inline src/preact.d.ts]: should this be in `DOMAttributed`?

File: src/preact.d.ts [hunk 3]
```diff
@@ -11,7 +11,8 @@ declare namespace preact {
...
+		ref?:<T,S>(el:Component<T,S>|Element)=>void;
```
Comment (developit) [inline src/preact.d.ts]: Ah, no you're right - sorry I missed that! Indeed, an Element VNode will only ever invoke its ref with either an `Element` or `null`.

File: src/preact.d.ts (hunk 2 above)
Comment (nojvek) [inline src/preact.d.ts]: Could someone answer why its called DomAttributed rather than DomAttributes? Its kind of confusing on the first look.

Comment (geppy) [inline src/preact.d.ts]: This typo was resolved in #402 (correctly, I think, based on usage).

**Pull request closed without merging**